File: projects/vpn-software/src/tun_device.rs
//! TUN device management for VPN tunnel
//!
//! Handles creation and management of TUN virtual network interfaces.

use std::io::{Read, Write};
use std::net::Ipv4Addr;

use crate::error::{Result, VpnError};

/// Configuration for TUN device
#[derive(Debug, Clone)]
pub struct TunConfig {
    /// Device name (e.g., "tun0")
    pub name: String,
    /// Local IP address
    pub address: Ipv4Addr,
    /// Network mask
    pub netmask: Ipv4Addr,
    /// MTU size
    pub mtu: u32,
}

impl Default for TunConfig {
    fn default() -> Self {
        Self {
            name: "tun0".to_string(),
            address: Ipv4Addr::new(10, 0, 0, 1),
            netmask: Ipv4Addr::new(255, 255, 255, 0),
            mtu: 1500,
        }
    }
}

/// Represents a TUN virtual network interface
pub struct TunDevice {
    /// Underlying TUN device
    device: tun::platform::Device,
    /// Device configuration
    config: TunConfig,
}

impl TunDevice {
    /// Create a new TUN device with the given configuration
    pub fn new(config: TunConfig) -> Result<Self> {
        let mut tun_config = tun::Configuration::default();
        tun_config
            .address(config.address)
            .netmask(config.netmask)
            .mtu(config.mtu as i32)
            .up();

        // Set device name if specified
        if !config.name.is_empty() {
            tun_config.name(&config.name);
        }

        let device = tun::create(&tun_config)
            .map_err(|e| VpnError::TunDeviceError(format!("Failed to create TUN device: {}", e)))?;

        Ok(Self { device, config })
    }

    /// Get the device configuration
    pub fn config(&self) -> &TunConfig {
        &self.config
    }

    /// Get the device name
    pub fn name(&self) -> &str {
        &self.config.name
    }

    /// Read a packet from the TUN device
    pub fn read_packet(&mut self) -> Result<Vec<u8>> {
        let mut buf = vec![0u8; self.config.mtu as usize];
        let n = self.device.read(&mut buf)
            .map_err(|e| VpnError::TunDeviceError(format!("Failed to read from TUN: {}", e)))?;
        buf.truncate(n);
        Ok(buf)
    }

    /// Write a packet to the TUN device
    pub fn write_packet(&mut self, packet: &[u8]) -> Result<()> {
        self.device.write_all(packet)
            .map_err(|e| VpnError::TunDeviceError(format!("Failed to write to TUN: {}", e)))?;
        Ok(())
    }

    /// Set the device IP address
    pub fn set_address(&mut self, address: Ipv4Addr) -> Result<()> {
        // Note: This would require platform-specific implementation
        self.config.address = address;
        Ok(())
    }

    /// Set the device MTU
    pub fn set_mtu(&mut self, mtu: u32) -> Result<()> {
        self.config.mtu = mtu;
        Ok(())
    }
}

/// Route management for VPN
pub struct RouteManager {
    /// VPN subnet
    vpn_subnet: Ipv4Addr,
    /// VPN netmask
    vpn_netmask: Ipv4Addr,
    /// Gateway address
    gateway: Ipv4Addr,
}

impl RouteManager {
    /// Create a new route manager
    pub fn new(vpn_subnet: Ipv4Addr, vpn_netmask: Ipv4Addr, gateway: Ipv4Addr) -> Self {
        Self {
            vpn_subnet,
            vpn_netmask,
            gateway,
        }
    }

    /// Add a route for the VPN subnet
    pub fn add_vpn_route(&self) -> Result<()> {
        // This would use platform-specific commands (ip route, route add, etc.)
        // For now, this is a placeholder
        tracing::info!("Adding route: {} via {}", self.vpn_subnet, self.gateway);
        Ok(())
    }

    /// Remove the VPN route
    pub fn remove_vpn_route(&self) -> Result<()> {
        tracing::info!("Removing route: {} via {}", self.vpn_subnet, self.gateway);
        Ok(())
    }

    /// Add a route for a specific peer
    pub fn add_peer_route(&self, peer_ip: Ipv4Addr) -> Result<()> {
        tracing::info!("Adding route for peer: {} via {}", peer_ip, self.gateway);
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tun_config_default() {
        let config = TunConfig::default();
        assert_eq!(config.name, "tun0");
        assert_eq!(config.address, Ipv4Addr::new(10, 0, 0, 1));
        assert_eq!(config.netmask, Ipv4Addr::new(255, 255, 255, 0));
        assert_eq!(config.mtu, 1500);
    }

    #[test]
    fn test_route_manager() {
        let manager = RouteManager::new(
            Ipv4Addr::new(10, 0, 0, 0),
            Ipv4Addr::new(255, 255, 255, 0),
            Ipv4Addr::new(10, 0, 0, 1),
        );

        // These would fail without root privileges, but we can test the creation
        assert_eq!(manager.vpn_subnet, Ipv4Addr::new(10, 0, 0, 0));
        assert_eq!(manager.gateway, Ipv4Addr::new(10, 0, 0, 1));
    }
}

File: projects/vpn-software/vpn/tun_device.py
"""
TUN device management for VPN tunnel.

Handles creation and management of TUN virtual network interfaces
on Linux. TUN devices operate at Layer 3 (IP packets).
"""

import os
import fcntl
import struct
import subprocess
import socket
from dataclasses import dataclass
from typing import Optional

from vpn.error import TunDeviceError

# Linux TUN constants
TUNSETIFF = 0x400454CA
IFF_TUN = 0x0001
IFF_NO_PI = 0x1000
IFF_TAP = 0x0002


@dataclass
class TunConfig:
    """TUN device configuration."""
    name: str = "tun0"
    address: str = "10.0.0.1"
    netmask: str = "255.255.255.0"
    mtu: int = 1500
    # Optional: peer address for point-to-point mode
    peer_address: Optional[str] = None


class TunDevice:
    """
    TUN virtual network interface.

    Creates and manages a TUN device for capturing/reading IP packets.
    Requires root privileges on Linux.
    """

    def __init__(self, config: TunConfig) -> None:
        self._config = config
        self._fd: Optional[int] = None
        self._name: Optional[str] = None
        self._opened = False

    @property
    def config(self) -> TunConfig:
        return self._config

    @property
    def name(self) -> Optional[str]:
        return self._name

    @property
    def is_open(self) -> bool:
        return self._opened

    def open(self) -> None:
        """Open and configure the TUN device."""
        try:
            # Open TUN clone device
            self._fd = os.open("/dev/net/tun", os.O_RDWR)

            # Prepare the ifreq struct
            # struct ifreq { char ifr_name[16]; unsigned short ifr_flags; ... }
            ifr_flags = IFF_TUN | IFF_NO_PI
            ifr = struct.pack("16sH", self._config.name.encode("utf-8"), ifr_flags)
            result = fcntl.ioctl(self._fd, TUNSETIFF, ifr)

            # Extract actual device name
            self._name = result[:16].rstrip(b"\x00").decode("utf-8")
            self._opened = True

            # Configure the interface
            self._configure_interface()

        except PermissionError:
            raise TunDeviceError(
                "Permission denied. TUN device requires root privileges."
            )
        except FileNotFoundError:
            raise TunDeviceError(
                "/dev/net/tun not found. Ensure TUN module is loaded."
            )
        except OSError as e:
            raise TunDeviceError(f"Failed to open TUN device: {e}")

    def _configure_interface(self) -> None:
        """Configure IP address, netmask, and MTU on the TUN device."""
        try:
            # Set IP address
            subprocess.run(
                ["ip", "addr", "add", f"{self._config.address}/{self._calc_prefix()}",
                 "dev", self._name],
                check=True,
                capture_output=True,
            )

            # Set MTU
            subprocess.run(
                ["ip", "link", "set", "dev", self._name, "mtu", str(self._config.mtu)],
                check=True,
                capture_output=True,
            )

            # Bring interface up
            subprocess.run(
                ["ip", "link", "set", "dev", self._name, "up"],
                check=True,
                capture_output=True,
            )

        except subprocess.CalledProcessError as e:
            raise TunDeviceError(f"Failed to configure TUN interface: {e.stderr.decode()}")

    def _calc_prefix(self) -> int:
        """Calculate CIDR prefix length from netmask."""
        mask_int = struct.unpack("!I", socket.inet_aton(self._config.netmask))[0]
        return bin(mask_int).count("1")

    def read_packet(self) -> bytes:
        """Read an IP packet from the TUN device."""
        if not self._opened or self._fd is None:
            raise TunDeviceError("TUN device not open")

        try:
            buf = os.read(self._fd, self._config.mtu + 64)
            return buf
        except OSError as e:
            raise TunDeviceError(f"Failed to read from TUN: {e}")

    def write_packet(self, packet: bytes) -> None:
        """Write an IP packet to the TUN device."""
        if not self._opened or self._fd is None:
            raise TunDeviceError("TUN device not open")

        try:
            os.write(self._fd, packet)
        except OSError as e:
            raise TunDeviceError(f"Failed to write to TUN: {e}")

    def close(self) -> None:
        """Close the TUN device."""
        if self._fd is not None:
            try:
                # Remove IP address
                subprocess.run(
                    ["ip", "addr", "del", f"{self._config.address}/{self._calc_prefix()}",
                     "dev", self._name],
                    capture_output=True,
                )
                # Bring interface down
                subprocess.run(
                    ["ip", "link", "set", "dev", self._name, "down"],
                    capture_output=True,
                )
            except Exception:
                pass
            os.close(self._fd)
            self._fd = None
            self._opened = False

    def __enter__(self) -> "TunDevice":
        self.open()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()


class RouteManager:
    """Manages system routing table for VPN traffic."""

    @staticmethod
    def add_route(destination: str, gateway: str, dev: Optional[str] = None) -> None:
        """Add a route to the system routing table."""
        cmd = ["ip", "route", "add", destination, "via", gateway]
        if dev:
            cmd.extend(["dev", dev])
        try:
            subprocess.run(cmd, check=True, capture_output=True)
        except subprocess.CalledProcessError as e:
            raise TunDeviceError(f"Failed to add route: {e.stderr.decode()}")

    @staticmethod
    def delete_route(destination: str) -> None:
        """Delete a route from the system routing table."""
        try:
            subprocess.run(
                ["ip", "route", "del", destination],
                check=True,
                capture_output=True,
            )
        except subprocess.CalledProcessError:
            pass  # Route may not exist

    @staticmethod
    def add_default_route(gateway: str, dev: str) -> None:
        """Add a default route through the VPN."""
        # Save the original default route
        try:
            result = subprocess.run(
                ["ip", "route", "show", "default"],
                capture_output=True, text=True,
            )
            original_default = result.stdout.strip()

            # Add specific route to VPN server through original gateway
            if original_default:
                subprocess.run(
                    ["ip", "route", "add"] + original_default.split(),
                    capture_output=True,
                )
        except Exception:
            pass

        # Add new default route through VPN
        subprocess.run(
            ["ip", "route", "add", "default", "via", gateway, "dev", dev],
            check=True, capture_output=True,
        )

    @staticmethod
    def restore_default_route() -> None:
        """Restore the original default route."""
        try:
            subprocess.run(
                ["ip", "route", "del", "default"],
                capture_output=True,
            )
        except Exception:
            pass
